### ISSUE TITLE
Benchmark against Brian 1

### ISSUE THREAD

**Issue by thesamovar** [brian-team/brian2#446]
Seems like B2 might be slower than B1 in some important cases?

**Comment by thesamovar** [comment]
This is still worth doing but at the moment it seems the most likely difference in terms of speed is that B2 has more Python overhead than B1 which makes it slower for small networks.

**Comment by thesamovar** [comment]
CNS meeting update: let's not do this systematically but at least check a few examples.

**Comment by thesamovar** [comment]
OK, the first plot is in, and it's not looking good for Brian 2:

![figure_1](https://cloud.githubusercontent.com/assets/2903870/11282841/655bce88-8ef9-11e5-8e7e-3c1b9dfecc73.png)

**Comment by mstimberg** [comment]
That's not looking great indeed. Could you give some more details about what we are seeing here? I guess the x axis is number of neurons? And is this all with codegen target weave?

**Comment by thesamovar** [comment]
Yeah it's just the same as the CUBA speed test example. The N is number of neurons, and there are 80 synapses per neuron. I haven't tried with different codegen targets yet so I don't actually know what they are using - will check that next. In the meantime, here is the same plot as above for a 10s run instead of a 1s run (showing that it's not just a startup time problem):

![figure_1](https://cloud.githubusercontent.com/assets/2903870/11285560/00538864-8f08-11e5-8364-707005e1bab0.png)

The script is in a new branch ``compare_to_brian_1``. The different times are as follows:

import brian: from the very start until ``from brian(2) import *`` has finished.
object creation: from the end of the previous step until just before run
initial run: a 1ms run
main run: a 1s/10s run
total: the whole time including all of the above

**Comment by thesamovar** [comment]
So it turned out that my Brian 1 installation was pure Python with no options set, so I set Brian 2 to numpy mode and of course now it does even worse in the comparison. I also changed the script so that it only has the NeuronGroup, and same problem. Indeed, in this case for the 'main run' part there is a constant difference between them, which (since they're log plots) means that there's a constant time factor, i.e. Brian 2 is N times slower than Brian 1 for some N>1. Anyway, looks like this ought to be diagnosable at least. Maybe something to do with the way we load arrays and write them back? I think because we were more restricted in what could happen in Brian 1 it might have been more efficient.

**Comment by thesamovar** [comment]
Also it is significantly slower, almost 10x for numpy target. For weave target on Brian 2 compared to numpy on Brian 1 (obviously unfair but for sanity checking purposes) you can see that it is converging to B2 being faster than B1 but it will take a long time because there is a very large overhead (either initial or per timestep - not sure yet). Definitely some work to do I think...

**Comment by mstimberg** [comment]
I did not yet have time to look at this in detail, but a few quick thoughts:
* It would be interesting to know whether this big difference is some kind of regression or whether it has been like this in all versions of Brian 2
* I wonder how much difference the symbolic vs. numerical approach to linear integration makes. In Brian 1 the matrix exponential is calculated from the numerical values of all constants whereas in Brian 2 we do it symbolically with the constant names.
* The initial run and main run comparison is not that fair to Brian 2, I think: In Brian 2, we do all the preparation again for the second run because we allow everything to change (i.e. you could change all constants, dt, etc. and it would still work) whereas in Brian 1 everything is fixed at creation time.

I think it is actually an important question whether we really need to allow this freedom of changing constants between run calls. It is a rarely used feature and (at least if it is not a constant that is used across objects) you can always use a `(shared, constant)` parameter in the equations to allow changes between runs more explicitly.

**Comment by thesamovar** [comment]
Do you have a tool for finding where something like this changes in the git history or do you just bisect by hand?

I changed both B1 and B2 to use Euler integration as an attempt to make it more comparable, and the differences are smaller now:

![figure_1](https://cloud.githubusercontent.com/assets/2903870/11298713/08431aaa-8f79-11e5-9333-ead000d7592d.png)

Still though, main run is taking about 2x longer for B2 than B1.

I think probably a lot of these constant overheads would be eliminated by joblib cacheing, but obviously that's not going to change per iteration overheads.

**Comment by thesamovar** [comment]
Another update. I modified the script so that it would store the time half way through the run using a network operation, and then compare the final time to this to give the 'half run time'. This should be a fair comparison for B1 and B2 right? I also noticed that it was ignoring the 'duration' parameter and always running for 1s, so these are 10s runs with the new half run time included:

![figure_1](https://cloud.githubusercontent.com/assets/2903870/11299446/b7f47372-8f7e-11e5-977b-d519142b3526.png)

Again, we still see that B1 is faster than B2 but now it's only about 50% faster (i.e. B2 takes 1.5x the time of B1).

**Comment by thesamovar** [comment]
And here's the same with refractory off:

![figure_1](https://cloud.githubusercontent.com/assets/2903870/11299793/89b8bd8a-8f81-11e5-98be-253d73cbe0ed.png)

The conclusion I'm tempted to draw is that we have a combination of problems. Firstly, I think we have much higher overheads than B1 and we should attempt to bring this down a bit. Secondly, refractoriness is slower on B2 than B1 (but probably only for the numpy target), but this is OK because we do it properly in B2 unlike B1 so the cost is worth it. There may also be inefficiencies in the numpy target for B2 that could be avoided. I'll start profiling in more detail now.

**Comment by mstimberg** [comment]
> Do you have a tool for finding where something like this changes in the git history or do you just bisect by hand?

I guess `git bisect` might come in handy :)

I still did not half the possibility to look at this more closely, but I agree that the "half-runtime" should be a fair measure of the actual runtime (but of course we also want to bring the Brian2-specific overheads down).

**Comment by mstimberg** [comment]
I had a quick look at a profiler run of the standard CUBA example and about 25% of the network run loop time seems to be spent in clock-related stuff. I did not verify this in detail yet, but I think I introduced a major inefficiency when we changed the clock system to store the time as a standard state variable. There are a few places where the clock time is accessed like a state variable (so with all the machinery that in principle allows the string-based indexing and so on) -- this is inefficient and will matter because we check the clock times very often during a run. This should be an easy thing to fix but will certainly only account for a part of the difference you were seeing above.

**Comment by thesamovar** [comment]
Ah, OK - well that would actually be a fairly substantial part of it then. The branch I created for this issue is based off the cython performance one, so we should either merge that now or put the clock fixes in a new branch based off master probably. Maybe prioritise reviewing and merging the cython performance one?

**Comment by mstimberg** [comment]
Agreed, I'll have a look at the Cython branch first before working on this issue.

**Comment by thesamovar** [comment]
Here are results for different versions of Brian 2 (1s runs):

Beta 4:

![beta4](https://cloud.githubusercontent.com/assets/2903870/11304501/2fb78400-8fa0-11e5-9db9-f249b7302eac.png)

Beta 3:

![beta3](https://cloud.githubusercontent.com/assets/2903870/11304503/33406222-8fa0-11e5-88fe-487a850e605b.png)

Beta 2:

![beta2](https://cloud.githubusercontent.com/assets/2903870/11304510/379f4dc4-8fa0-11e5-9a14-adbc627a34bf.png)

So it looks like this is a fairly recent performance regression: in all these versions the half run time is the same for B1 and B2.

**Comment by thesamovar** [comment]
This is current master:

![master](https://cloud.githubusercontent.com/assets/2903870/11304634/dee6983a-8fa0-11e5-9bd4-e9cab6f4a263.png)

The runtime problem is there, but note that the initial run performance is better in current master than in either the branch for this issue (based on cython performance) or the earlier beta releases. So apparently we fixed some initial run performance between b4 and master, but cython performance made it worse again. I suspect that the optimisation stuff actually contributes to this problem: it's quite a heavy computation (but can be cached).

**Comment by thesamovar** [comment]
Git bisect is amazing! Good call. ``cd384495f71d0ae1c137bcac499ebf83e694bfa8 is the first bad commit``, which is when we converted ``Clock`` into ``Group``. The commit before that the speed is just as fast for Brian 1 and 2, so it seems this is indeed the whole problem (at least for CUBA without refractoriness). I'll wait for you to fix this one I think.

**Comment by thesamovar** [comment]
Also I just realised another place where we're not going to do well in Brian 2 compared to Brian 1: ``Synapses.connect``. That's because at the moment we're forced to do N^2 operations. Still, if we can at least get comparable/better runtime performance on B2 compared to B1 that would be a great start. We'll see about fixing construction of synapses later.

**Comment by thesamovar** [comment]
OK, with the latest changes merged in, here's how it's looking now.

For Euler integration, no refractoriness or synapses, B1 and B2 are looking about the same now whether or not spike monitoring is on. With refractoriness on, B2 is a little slower for numpy, but I think it's a reasonable price to pay for the refractoriness being more correctly evaluated (and this might be fixed with CSE and boolean simplifications which will be done later).

Synapses are slower than Connection. This might be worth looking into actually, since now for ``v+=w`` it should be translated to ``add.at`` which should be fast I would have thought.

With exact integration on, it's a lot slower, and this is a known problem that we're investigating elsewhere.

Now, here's a comparison of the simplest case (neurons only, no refractoriness, euler integration, no spike monitor) with default Brian 1 (pure Python) versus default Brian 2 with various codegen targets (all with the Cython spike queue):

![compare_to_brian1](https://cloud.githubusercontent.com/assets/2903870/11511521/847a3b58-9861-11e5-8a55-ccd44bc977be.png)

This is interesting because it probably represents the typical Brian 1 to Brian 2 experience, as the majority of users probably didn't investigate all the codegen options available in B1. What's clear is that at least in this case, B2 is winning, very much so in the case of large N. It's also the case that Cython appears to be much better than weave because of lower overheads.

Possibly we should switch the default search order for codegen targets to cython, weave, numpy?

Putting the monitor on is a similar picture, but at low N B2 appears to be a little slower than B1 for numpy and weave:

![compare_to_brian1](https://cloud.githubusercontent.com/assets/2903870/11511672/58e29caa-9862-11e5-9c29-20ac28d109d8.png)

Here's with refractoriness on:

![compare_to_brian1](https://cloud.githubusercontent.com/assets/2903870/11511806/f7757efa-9862-11e5-8138-4415b90e9658.png)

Note that again, B2 a little slower than B1 with numpy, but always faster with Cython and usually faster with weave. It does look as though Cython and weave might be crossing over for very large N though.

Here's with synapses on:

![compare_to_brian1](https://cloud.githubusercontent.com/assets/2903870/11512051/30a154dc-9864-11e5-968a-d729b7532111.png)

For very large N, B2/weave and B2/cython are faster for runtime and cython faster than weave, but B2/numpy is slower. For small N, B2 is always slower than B1. Unsurprisingly (given the O(N^2) factor), for the initial object creation + initial run, B2 is fairly slow.

It sort of looks like the poor runtime performance for B2 for lowish numbers of synapses is probably due to overheads that are worst for weave, a little better for cython, better for numpy (but much better for B1). This could be another thing we look into.

Finally, for exact integration (I ran it for 10s rather than 1s because it seemed rather noisy, so the relative contribution of initial computation to ongoing computation is different):

![compare_to_brian1](https://cloud.githubusercontent.com/assets/2903870/11512336/86ae6e86-9865-11e5-9c48-a57590d37eb5.png)

Here again it looks like Cython on B2 is the winner, with a large gap between it and B2/weave.

So conclusions / discussion points:

* Obvious Python overheads seem not to be an issue in B2 given that Cython performance is very good. We can probably therefore close #447? Or should we investigate some more examples before we do this.
* There are some overheads issues with weave. Can we solve these? Unclear. Maybe switching B2 to prefer Cython to weave is a good idea?
* There are probably some runtime inefficiencies for synapses that mean that it runs a lot slower than Connection used to. We should check how much of this is due to the spike queue (which Connection didn't have). If it's a lot, perhaps it's worth introducing a no-delay / homogeneous-delay alternative spike queue? Actually, did we ever correctly implement the optimisation for the homogeneous case? I remember we noticed the problem with it before and discussed solutions, but I don't remember if we actually did it.
* There's still the O(N^2) synapses problem, and the slow refractoriness/exact integration issues for B2/numpy to deal with: both of these are separate issues though.

**Comment by thesamovar** [comment]
Ignore the synapses one, I had the python SpikeQueue for some reason. With the Cython SpikeQueue, here are the results (for 10s run to reduce noise):

![compare_to_brian1](https://cloud.githubusercontent.com/assets/2903870/11513801/11f8d812-986d-11e5-821c-2225ac7e1a97.png)

Now this is more consistent with the others. Cython is the fastest, with weave having some overhead issues again. Now, runtime performance of Cython at least is better for B2 than B1.

Looking into the runtime overhead a bit more, I think it can be reduced by about a third. This is part of ``SynapticPathway``:

```
    def push_spikes(self):
        # Push new events (e.g. spikes) into the queue
        events = self.eventspace[:self.eventspace[len(self.eventspace)-1]]
        if len(events):
            self.queue.push(events)
        # Get the spikes
        spiking_synapses = self.queue.peek()
        spiking_synapses_var = self.variables['spiking_synapses']
        spiking_synapses_var.resize(len(spiking_synapses))
        spiking_synapses_var.set_value(spiking_synapses)
        # Advance the spike queue
        self.queue.advance()
```

It gets called every time step. If we replace this with the following code (which I think gives the same results but is obviously a hack):

```
    def push_spikes(self):
        # Push new events (e.g. spikes) into the queue
        events = self.eventspace[:self.eventspace[len(self.eventspace)-1]]
        if len(events):
            self.queue.push(events)
        # Get the spikes
        spiking_synapses = self.queue.peek()
        spiking_synapses_var = self.variables['spiking_synapses']
        spiking_synapses_var.data = spiking_synapses_var._data =  spiking_synapses
        spiking_synapses_var.shape = spiking_synapses.shape
        # Advance the spike queue
        self.queue.advance()
```

Then we get a reduction by a third in the time. Removing everything in this method entirely reduces the time by a half, meaning that two thirds of the time spent in this function (amounting to one third of the total time spent on synapses) is wasted resizing and copying into a dynamic array. For the CUBA example with 4k neurons, the synapses at the moment are taking about two thirds of the total runtime, so with this hack/optimisation the total runs about 28% faster.

I guess this is implemented like this so that standalone and runtime have the same mechanism. Perhaps we should introduce either a "dynamic reference array" or just have a different mechanism for runtime than standalone? I believe that none of the runtime templates make any use of the fact that this is a dynamic array so this should be possible.

**Comment by thesamovar** [comment]
For the weave overhead, I have a solution based on looking at the weave source code. Essentially, you can skip some namespace, stack and cache code and directly call the function that ends up getting called anyway, and after you do that the difference between cython and weave disappears (and actually weave gets a little faster in some cases). Here is an example where weave gets faster (refractoriness on):

![Uploading compare_to_brian1.png…]()

I'm just running the new code through the tests and if it passes I'll push it to this branch. I think this 'hack' ought to be safe: weave's source code hasn't changed in many years and it's unlikely to do so in the future.

**Comment by thesamovar** [comment]
Ah, maybe hit post too soon:

![compare_to_brian1](https://cloud.githubusercontent.com/assets/2903870/11516982/b36bc4a0-987e-11e5-945d-06b0001c7a87.png)

**Comment by thesamovar** [comment]
Tests pass, but I'll make a PR now to run the full suite of tests. Also, here's the new weave code since I forgot this was an issue not a PR so you won't see the diff in the thread:

```
    def run(self):
        if self.compiled_python_pre is not None:
            exec self.compiled_python_pre in self.python_code_namespace
        if self._done_first_run:
            ret_val = self._compiled_func(self.namespace, {})
        else:
            self._inline_args = (self.annotated_code, self.namespace.keys())
            self._inline_kwds = dict(
                local_dict=self.namespace,
                support_code=self.code.support_code,
                compiler=self.compiler,
                headers=self.headers,
                define_macros=self.define_macros,
                libraries=self.libraries,
                extra_compile_args=self.extra_compile_args,
                extra_link_args=self.extra_link_args,
                include_dirs=self.include_dirs,
                library_dirs=self.library_dirs,
                verbose=0)
            with std_silent():
                ret_val = weave.inline(*self._inline_args, **self._inline_kwds)
            self._compiled_func = function_cache[self.annotated_code]
            self._done_first_run = True
        if self.compiled_python_post is not None:
            exec self.compiled_python_post in self.python_code_namespace
        return ret_val
```

**Comment by mstimberg** [comment]
Ah great, I had noticed before that the weave call had quite a bit overhead but I never looked into the details. It's now no longer an issue but in general I wouldn't mind to default to cython before weave. However, don't we have much longer compilation times with cython compared to weave? I guess your tests were all loading the compiled code from disk? The compilation time is still something that bothers me in general, for first-time users that try out a trivial example there's always a pause of some seconds before the simulation starts that makes Brian look slow. But I don't see any way around it, we don't want to default to `numpy` either, your numbers quite clearly show the advantage of using `weave`/`cython` instead.

About the `spiking_synapses` issue: there was probably a good reason for doing it this way, but I don't remember... It seems the whole problem arises only because we get the spiking synapses in the `synapses_push` template and we need to transfer it over to the `synapses` template. This is actually different from standalone, where we do only advance + push in the `synapses_push` template and do the `peek` in `synapses` directly. I don't really see why we can't do the same in runtime, this would make the whole issue go away and probably make things even faster than your hacky solution (you access attributes of a Python object).

**Comment by mstimberg** [comment]
Well, what I said above about `spiking_synapses` did not make that much sense... First of all I was confused about your "access of attributes of a Python object", since of course this is Python code not Cython/weave... And calling `peek` directly in the `synapses` template is of course trickier than in standalone, since it returns a Python object. Nevertheless, it is feasible, but it needs some change in the general namespace mechanism since we have to give a reference to the `SpikeQueue` object (which is not represented by a `Variable`) to weave. If I hack this in, then we can get the spikes in the `synapses` template like this:
```C++
// Get the spikes
const PyArrayObject *_spiking_synapses_obj = (PyArrayObject *)PyObject_CallMethod(_queue, "peek", "");
const int *_spiking_synapses = (int *)_spiking_synapses_obj->data;
const int _num_spiking_synapses = _spiking_synapses_obj->dimensions[0];
```
This should be made a bit safer with respect to integer types and so on, but in general it should work. I wonder whether we cannot even directly store a pointer to the underlying C++ `SpikeQueue` (if using the Cython spikequeue), so that we don't have to call the Python method at all.

Either way, this change has a big effect (as predicted from your changes), on my machine it reduces the runtime of `synapses_pre` and `synapses_1_pre` to 60% of its previous values under weave. This seems to be important enough to do a few somewhat fundamental changes to allow the direct access. I'll try to have a go at this right now, if it is not interfering with something you were doing.

**Comment by thesamovar** [comment]
Go for it, I'm busy on other stuff all afternoon.

**Comment by mstimberg** [comment]
I submitted the change to the branch, it seems to work fine. I did not find an obvious way to interact with the C++ `SpikeQueue` directly (I would have to access the `thisptr` in the Cython `SpikeQueue` somehow).
I also had a quick look at doing the push completely in templates as well, but this is more difficult (we have to allocate copy memory and so on) and a quick and dirty solution did not seem to give any performance improvement, so I did not further pursue this.
It is kind of satisfying that the profiling times in the CUBA example switched from using seconds to milliseconds as the unit on my machine now :) It is kind of annoying that the `SpikeMonitor` is now the most expensive codeobject, though... It might be worth having another less elegant but faster solution there directly in the template.
